### PR TITLE
DW-2513 send files once

### PR DESCRIPTION
**Pull request by mark-r-m**
Write a flag file back to S3 when a collection has finished sending to NiFi. on a re-run, skip collection files that were previously finished.

### CODE REVIEW

File: src/main/kotlin/app/batch/S3StatusFileWriter.kt [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+class S3StatusFileWriter : S3Utils() {
+
+    fun writeStatus(originalS3Key: String) {
+
+        val payloadText = "Finished $originalS3Key"
+        val payloadBytes = payloadText.toByteArray(Charsets.UTF_8)
+        val payloadInputStream: InputStream = ByteArrayInputStream(payloadBytes)
+
+        // i.e. sourceFile: s3://bucket/business-data-export/JobNumber/1990-01-31/db.user.data-0001.bz2.enc
+        // i.e. statusFile: s3://bucket/business-sender-status/JobNumber/1990-01-31/db.user.data-0001.bz2.enc.finished
```
Comment (dan-hill2802) [inline src/main/kotlin/app/batch/S3StatusFileWriter.kt:25]: These might be useful in the readme

File: src/main/kotlin/app/batch/S3StatusFileWriter.kt [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+            e.printStackTrace()
+            logger.error("SdkClientException processing '$originalS3Key': ${e.message}")
+        }
+    }
+
+    companion object {
+        val logger: Logger = LoggerFactory.getLogger(S3StatusFileWriter::class.toString())
+    }
+
+}
```
Comment (dan-hill2802) [inline src/main/kotlin/app/batch/S3StatusFileWriter.kt]: New line please

File: src/main/kotlin/app/batch/S3Utils.kt [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+
+    fun prefixFolder(): String {
+        return if (s3PrefixFolder.endsWith("/")) {
+            s3PrefixFolder
+        }
+        else {
+            "$s3PrefixFolder/"
+        }
+    }
+}
```
Comment (dan-hill2802) [inline src/main/kotlin/app/batch/S3Utils.kt]: New line please

File: src/main/kotlin/app/domain/DomainClasses.kt [hunk 4]
```diff
@@ -7,6 +7,6 @@ data class EncryptionMetadata (val initializationVector: String,
                                val cipherText: String,
                                var plaintext: String)
 
-data class EncryptedStream(val inputStream: InputStream, val fileName: String, val encryptionMetadata: EncryptionMetadata)
+data class EncryptedStream(val inputStream: InputStream, val fileName: String, val fullPath: String, val encryptionMetadata: EncryptionMetadata)
 data class DataKeyResult(val dataKeyEncryptionKeyId: String, val plaintextDataKey: String, val ciphertextDataKey: String)
-data class DecryptedStream(val inputStream: InputStream, val filename: String)
\ No newline at end of file
+data class DecryptedStream(val inputStream: InputStream, val fileName: String, val fullPath: String)
```
Comment (dan-hill2802) [inline src/main/kotlin/app/domain/DomainClasses.kt]: New line

File: src/test/kotlin/app/TestUtils.kt [hunk 5]
```diff
@@ -0,0 +1,14 @@
...
+
+class TestUtils {
+
+    companion object {
+
+        fun once(): VerificationMode? {
+            return Mockito.times(1)
+        }
+    }
+}
```
Comment (dan-hill2802) [inline src/test/kotlin/app/TestUtils.kt]: New line

File: src/test/kotlin/app/batch/HttpWriterTest.kt [hunk 6]
```diff
@@ -1,121 +1,200 @@
...
+        assertEquals("Successfully posted 'exporter-output/job01/db.core.addressDeclaration-000001.txt.bz2.enc': response '200'", formattedMessages[4])
     }
 
     @Test
-    fun testAlternateFilenamePattern() {
-        val byteArray = "hello, world".toByteArray()
+    fun test_will_write_to_nifi_when_valid_file_with_embedded_hyphens_in_dbname() {
         val filename = "db.core-with-hyphen.addressDeclaration-000001.txt.bz2.enc"
-        val decryptedStream = DecryptedStream(ByteArrayInputStream(byteArray), filename)
+        val decryptedStream = DecryptedStream(ByteArrayInputStream(byteArray), filename, "$s3Path/$filename")
```
Comment (dan-hill2802) [inline src/test/kotlin/app/batch/HttpWriterTest.kt:107]: seems strange that filename is here twice. 
Is this the definition: `data class DecryptedStream(val inputStream: InputStream, val filename: String)`?
This occurs in several places, but I won't comment those

File: src/test/kotlin/app/services/impl/HttpKeyServiceTest.kt [hunk 7]
```diff
@@ -186,7 +184,4 @@ class HttpKeyServiceTest {
     @Autowired
     private lateinit var httpClientProvider: HttpClientProvider
 
-    companion object {
-        val logger: Logger = LoggerFactory.getLogger(HttpKeyService::class.toString())
-    }
 }
```
Comment (dan-hill2802) [inline src/test/kotlin/app/services/impl/HttpKeyServiceTest.kt:187]: new line

**Review by dan-hill2802 — CHANGES_REQUESTED** [review]

File: src/main/kotlin/app/batch/S3StatusFileWriter.kt (hunk 1 above)
Comment (mark-r-m) [reply src/main/kotlin/app/batch/S3StatusFileWriter.kt:25]: Yes, it needs a proper overhaul.
If it's ok I'll do that in the Integration Tests PR?

File: src/test/kotlin/app/batch/HttpWriterTest.kt (hunk 6 above)
Comment (mark-r-m) [reply src/test/kotlin/app/batch/HttpWriterTest.kt:107]: We need the filename. (for nifi) and the full path (for the S3 status writer)

Comment (mark-r-m) [reply src/test/kotlin/app/batch/HttpWriterTest.kt:107]: That definition is from master :-)

**Review by dan-hill2802 — APPROVED** [review]

**Comment by dan-hill2802** [comment]
Given the size and scope of this change I recommend an additional approval

**Pull request merged**